### PR TITLE
show User's status in UserProfileModal + link to their current location

### PR DESCRIPTION
**Pull request by oleksii-front**
closes: https://github.com/sparkletown/internal-sparkle-issues/issues/769

DO NOT MERGE in `env/github`

### CODE REVIEW

**Review by yurii-lubynets — DISMISSED** [review]

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 1]
```diff
@@ -55,72 +60,96 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
...
+                    }}
+                  />
+                </div>
+                <div className="profile-text">
+                  <h2 className="italic">
+                    {selectedUserProfile.partyName || DEFAULT_PARTY_NAME}
+                  </h2>
+                  {`is ${status} in `}
+                  <span className="profile-text-venue-name" onClick={goToVenue}>
+                    {selectVenue}
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx:115]: Should we be showing the venue name here, rather than the `venueId`?

Looking at the designs, it looks like we should be.

**Review by 0xdevalias — CHANGES_REQUESTED** [review]
Questions/comments/changes.

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 2]
```diff
@@ -38,6 +41,8 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
   const { user } = useUser();
 
   const { selectRecipientChat } = useChatSidebarControls();
+  const { recentWorldUsers } = useRecentWorldUsers();
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx:44]: @sparkletown/product @sparkletown/design @mike-lvov Also, instead of just `online` / `offline`, should we be showing any/all of their status types here?

File: src/components/organisms/UserProfileModal/UserProfileModal.scss [hunk 3]
```diff
@@ -62,6 +62,15 @@ $modal-max-width: 540px;
...
       grid-column: 2;
       grid-row: 1;
+
+      .profile-text-venue-name {
+        text-decoration: underline;
+        cursor: pointer;
+
+        &:hover {
+          color: $yellow;
+        }
```
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.scss]: https://github.com/sparkletown/sparkle/pull/1484/commits/b8236031f041d56c96fe29bbe98acb049df854cb

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 4]
```diff
@@ -55,72 +60,96 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
...
-                      RANDOM_AVATARS[
-                        Math.floor(
-                          chosenUserId.charCodeAt(0) % RANDOM_AVATARS.length
-                        )
-                      ];
-                  }}
-                />
+        {isShown ? (
+          <LoadingPage />
+        ) : (
```
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/e9391f6ee038146c4393c04364ee855e59fc84c4

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/c91da9a9f91b044f1211459a072d3dc813ecdf54

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 5]
```diff
@@ -55,72 +60,96 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
     closeUserProfileModal();
   }, [selectRecipientChat, closeUserProfileModal, chosenUserId]);
 
+  const status = useMemo(() => {
+    const isOnline = recentWorldUsers.find(
+      (worldUser) => worldUser.id === chosenUserId
+    );
+    return isOnline ? "online" : "offline";
+  }, [chosenUserId, recentWorldUsers]);
+
```
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/c91da9a9f91b044f1211459a072d3dc813ecdf54

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/1087979e5c2384770f952676d4be29cdbc738d80

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 6]
```diff
@@ -55,72 +60,96 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
...
+  const selectVenue = useMemo(() => {
+    return selectedUserProfile?.enteredVenueIds?.[0] || "";
+  }, [selectedUserProfile]);
+
+  const goToVenue = useCallback(() => {
+    show();
+    enterVenue(selectVenue);
+  }, [selectVenue, show]);
```
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/1087979e5c2384770f952676d4be29cdbc738d80

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 7]
```diff
@@ -55,72 +60,96 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
...
+  const status = useMemo(() => {
+    const isOnline = recentWorldUsers.find(
+      (worldUser) => worldUser.id === chosenUserId
+    );
+    return isOnline ? "online" : "offline";
+  }, [chosenUserId, recentWorldUsers]);
+
+  const selectVenue = useMemo(() => {
+    return selectedUserProfile?.enteredVenueIds?.[0] || "";
+  }, [selectedUserProfile]);
```
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/1087979e5c2384770f952676d4be29cdbc738d80

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 8]
```diff
@@ -38,6 +41,8 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
   const { user } = useUser();
 
   const { selectRecipientChat } = useChatSidebarControls();
+  const { recentWorldUsers } = useRecentWorldUsers();
+  const { isShown, show } = useShowHide();
```
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/e9391f6ee038146c4393c04364ee855e59fc84c4

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 9]
```diff
@@ -55,72 +60,96 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
...
+              <div className="profile-extras">
+                {venue?.profile_questions?.map((question) => (
+                  <React.Fragment key="question.text">
+                    <p className="light question">{question.text}</p>
+                    <h6>
+                      {/*
+                        @debt typing - need to support known User interface with unknown question keys
+                        @ts-ignore */}
+                      {selectedUserProfile[question.name] || //@debt typing - look at the changelog, was this a bug?
+                        DEFAULT_EDIT_PROFILE_TEXT}
```
[the remark this answers is not in the document]
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: Not in this PR

File: src/components/organisms/UserProfileModal/UserProfileModal.scss [hunk 10]
```diff
@@ -62,6 +62,15 @@ $modal-max-width: 540px;
     .profile-text {
       grid-column: 2;
       grid-row: 1;
+
+      &__venueId {
```
Comment (mike-lvov) [inline src/components/organisms/UserProfileModal/UserProfileModal.scss:66]:
```suggestion
      &__recent-venue {
```

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 11]
```diff
@@ -45,7 +46,11 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
     closeUserProfileModal,
   } = useProfileModalControls();
 
-  const chosenUserId = selectedUserProfile?.id;
+  const chosenUserId = selectedUserProfile?.id ?? "";
```
Comment (mike-lvov) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:49]:
```suggestion
  const chosenUserId = selectedUserProfile?.id;
```

Any reason to have "" as default? It can lead to unexpected behaviours

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 12]
```diff
@@ -55,6 +60,12 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
     closeUserProfileModal();
   }, [selectRecipientChat, closeUserProfileModal, chosenUserId]);
 
+  const goToVenue = useCallback(() => {
```
Comment (mike-lvov) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:63]:
```suggestion
  const enterRecentUserVenue = useCallback(() => {
```

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 13]
```diff
@@ -86,6 +97,10 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
                 <h2 className="italic">
                   {selectedUserProfile.partyName || DEFAULT_PARTY_NAME}
                 </h2>
+                {`is ${status} in `}
+                <span className="profile-text__venueId" onClick={goToVenue}>
+                  {selectedUserLastSeenInVenueId ?? ""}
```
Comment (mike-lvov) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:102]:
```suggestion
                  {selectedUserLastSeenInVenueId}
```

File: src/hooks/users.ts [hunk 14]
```diff
@@ -105,6 +105,31 @@ export const useRecentWorldUsers = (): {
...
+  const { recentWorldUsers } = useRecentWorldUsers();
+
+  const user = useMemo(() => {
+    return recentWorldUsers.find((worldUser) => worldUser.id === userId);
+  }, [userId, recentWorldUsers]);
+
+  const userLastSeenIn = Object.keys(user?.lastSeenIn ?? {})[0];
+
+  return { user, userLastSeenIn };
+};
```
Comment (mike-lvov) [inline src/hooks/users.ts:123]: What's the point of having this hook, if you have `selectedUserProfile` at the level of calling it?

File: src/hooks/users.ts [hunk 15]
```diff
@@ -105,6 +105,31 @@ export const useRecentWorldUsers = (): {
...
+
+export const useRecentUserStatus = (userId: string): RecentUserStatusType => {
+  const { user } = useRecentWorldUser(userId);
+
+  if (user?.status) return RecentUserStatusType.busy;
+
+  return user ? RecentUserStatusType.online : RecentUserStatusType.offline;
+};
```
Comment (mike-lvov) [inline src/hooks/users.ts:131]: this should be implemented as a util function. That function will accept `user` obj. `getUserDisplayStatus`

File: src/hooks/users.ts (hunk 14 above)
Comment (mike-lvov) [reply src/hooks/users.ts:123]: Ah, I guess to check if the user is online and return his location. Well, it should be divided into two. First hook should check if user is online. 
NOTE: Also reuse it in `UserAvatar`

image

Second util should be util function and return either undefined or location. By the way. `getUserCurrentLocation`. It will return either a string or undefined

Something similar to this
`const [locationName] = Object.keys(profileLocationData);`

**Review by mike-lvov — CHANGES_REQUESTED** [review]
First pass

File: src/components/organisms/UserProfileModal/UserProfileModal.scss (hunk 10 above)
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.scss:66]: Why. It contains venue id. I think that is right name for the block.

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 11 above)
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx:49]: you're right. I need to hide it when it is not defined

File: src/hooks/users.ts (hunk 14 above)
Comment (oleksii-front) [reply src/hooks/users.ts:123]: https://github.com/sparkletown/sparkle/pull/1484#discussion_r643612367

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 11 above)
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/6fb7c069dc5c0c57ab4df3f0cc0c358dfe307793

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 12 above)
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/68aadba523f084402f3810d1af68d9c32328e498

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 13 above)
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: https://github.com/sparkletown/sparkle/pull/1484/commits/457e5feeaa117f4b5364121ce5a420d19083cfcc

File: src/components/organisms/UserProfileModal/UserProfileModal.scss (hunk 10 above)
Comment (oleksii-front) [reply src/components/organisms/UserProfileModal/UserProfileModal.scss]: https://github.com/sparkletown/sparkle/pull/1484/commits/315d06f92e49515c4b4f1c6603384c2e37eba7e7

File: src/hooks/users.ts (hunk 14 above)
Comment (oleksii-front) [reply src/hooks/users.ts]: https://github.com/sparkletown/sparkle/pull/1484/commits/7b918f3a1e72d5576ebab3326949b537d1088345

File: src/hooks/users.ts (hunk 15 above)
Comment (oleksii-front) [reply src/hooks/users.ts]: https://github.com/sparkletown/sparkle/pull/1484/commits/0e4e679fca9153d9b74759272963562f5f6903fc

**Comment by yurii-lubynets** [comment]
@sparkletown/engineering kindly ask you to review this PR. Actually the core changes were made by @oleksii-front - I'm updating the PR and fix issues

File: src/hooks/users.ts [hunk 16]
```diff
@@ -105,6 +108,40 @@ export const useRecentWorldUsers = (): {
   );
 };
 
+type LastSeenAt = {
```
Comment (0xdevalias) [inline src/hooks/users.ts:111]:
```suggestion
export interface LastSeenAt {
```

File: src/hooks/users.ts [hunk 17]
```diff
@@ -1,9 +1,12 @@
 import { useMemo } from "react";
+import { createUrlSafeName } from "api/admin";
 
-import { User } from "types/User";
+import { User, RecentUserStatusType } from "types/User";
 
 import { WithId } from "utils/id";
 import { normalizeTimestampToMilliseconds } from "utils/time";
+import { getUserCurrentLocation } from "utils/profile";
+import { getUserDisplayStatus } from "utils/profile";
```
Comment (0xdevalias) [inline src/hooks/users.ts:9]:
```suggestion
import { getUserCurrentLocation, getUserDisplayStatus } from "utils/profile";
```

File: src/hooks/users.ts [hunk 18]
```diff
@@ -105,6 +108,40 @@ export const useRecentWorldUsers = (): {
...
+  const venueName = getUserCurrentLocation(user) ?? "";
+  const venueUrl = createUrlSafeName(venueName);
+
+  const lastSeenAt: LastSeenAt = {
+    venueName,
+    venueUrl,
+  };
+
+  return { user, lastSeenAt };
+};
```
Comment (0xdevalias) [inline src/hooks/users.ts:137]: We have `useWorldUsersById` (and the typing fixed version `useWorldUsersByIdWorkaround`) that will be more efficient to use than searching through `useRecentWorldUsers` with `.find`

File: src/hooks/users.ts [hunk 19]
```diff
@@ -105,6 +108,40 @@ export const useRecentWorldUsers = (): {
...
+    return recentWorldUsers.find((worldUser) => worldUser.id === userId);
+  }, [userId, recentWorldUsers]);
+
+  const venueName = getUserCurrentLocation(user) ?? "";
+  const venueUrl = createUrlSafeName(venueName);
+
+  const lastSeenAt: LastSeenAt = {
+    venueName,
+    venueUrl,
+  };
```
Comment (0xdevalias) [inline src/hooks/users.ts:134]: We shouldn't 'overload' the term 'last seen at' since we explicitly use that term to relate to a different concept already (`lastSeenAt` is the time the user was last seen, regardless of venue/etc)

Since both `lastSeenIn` and `lastSeenAt` can both be read directly off the `user`, do we need to be doing anything extra here?

File: src/hooks/users.ts [hunk 20]
```diff
@@ -105,6 +108,40 @@ export const useRecentWorldUsers = (): {
...
+
+  return { user, lastSeenAt };
+};
+
+export const useRecentUserStatus = (userId?: string): RecentUserStatusType => {
+  const { user } = useRecentWorldUser(userId);
+
+  return getUserDisplayStatus(user);
+};
```
Comment (0xdevalias) [inline src/hooks/users.ts:143]: This probably doesn't need to be a custom hook, we can just call the `getUserDisplayStatus` within the component/etc that needs it I expect.

File: src/utils/profile.ts [hunk 21]
```diff
@@ -19,5 +20,21 @@ export const updateProfileEnteredVenueIds = async (
   }
 };
 
+export const getUserDisplayStatus = (
+  user?: WithId<User>
+): RecentUserStatusType => {
+  if (user?.status) return RecentUserStatusType.busy;
+
+  return user ? RecentUserStatusType.online : RecentUserStatusType.offline;
+};
```
Comment (0xdevalias) [inline src/utils/profile.ts:29]: The logic here seems weird/wrong?

When `user?.status` is true we'll never get to the 2nd line.

File: src/utils/profile.ts [hunk 22]
```diff
@@ -19,5 +20,21 @@ export const updateProfileEnteredVenueIds = async (
...
 export const isCompleteProfile = (profile: FirebaseReducer.Profile<User>) =>
   isTruthy(profile.partyName) && isTruthy(profile.pictureUrl);
+
+export const getUserCurrentLocation = (user?: WithId<User>) => {
+  if (!user?.lastSeenIn) {
+    return undefined;
+  }
+
+  return Object.keys(user.lastSeenIn)[0];
+};
```
Comment (0xdevalias) [inline src/utils/profile.ts:40]: What's the intent of this?

File: src/hooks/users.ts [hunk 23]
```diff
@@ -105,6 +108,40 @@ export const useRecentWorldUsers = (): {
...
+  lastSeenAt: LastSeenAt;
+} => {
+  const { recentWorldUsers } = useRecentWorldUsers();
+
+  const user = useMemo(() => {
+    return recentWorldUsers.find((worldUser) => worldUser.id === userId);
+  }, [userId, recentWorldUsers]);
+
+  const venueName = getUserCurrentLocation(user) ?? "";
+  const venueUrl = createUrlSafeName(venueName);
```
Comment (0xdevalias) [inline src/hooks/users.ts:129]: What does `createUrlSafeName` do? Not sure if this is going to be correct. `venueName` is something entirely different to `venueId`. The URLs use `venueId`.

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 24]
```diff
@@ -127,6 +134,13 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
                 <h2 className="italic">
                   {selectedUserProfile.partyName || DEFAULT_PARTY_NAME}
                 </h2>
+                {renderStatus}
+                <a
+                  href={lastSeenAt.venueUrl}
+                  className="profile-text__recent-venue"
+                >
+                  {lastSeenAt.venueName}
+                </a>
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:143]: I believe we have helpers in `utils/url` for getting the correct venue URL/etc. We can probably just make use of those inline here?

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 25]
```diff
@@ -45,17 +46,23 @@ export const UserProfileModal: React.FC<UserProfileModalProps> = ({
...
+  const { lastSeenAt } = useRecentWorldUser(chosenUserId);
+  const status = useRecentUserStatus(chosenUserId);
 
   const profileQuestions = venue?.profile_questions;
 
   const openChosenUserChat = useCallback(() => {
     if (!chosenUserId) return;
 
     selectRecipientChat(chosenUserId);
-    // NOTE: Hide the modal, after the chat is opened;
```
Comment (0xdevalias) [inline src/components/organisms/UserProfileModal/UserProfileModal.tsx:55]: Why was this removed from here?

**Review by 0xdevalias — CHANGES_REQUESTED** [review]
Questions/comments/changes/etc. Quick review, not super deep/thorough.

File: src/hooks/users.ts (hunk 18 above)
Comment (yurii-lubynets) [reply src/hooks/users.ts]: got it. fixed in 5a3581bc4abfb14880e2e621a12c0994e01cace8

File: src/hooks/users.ts (hunk 19 above)
Comment (yurii-lubynets) [reply src/hooks/users.ts]: definitely agree! fixed in 5a3581bc4abfb14880e2e621a12c0994e01cace8

File: src/hooks/users.ts (hunk 20 above)
Comment (yurii-lubynets) [reply src/hooks/users.ts]: updated this hook. fixed in 5a3581bc4abfb14880e2e621a12c0994e01cace8

File: src/utils/profile.ts (hunk 21 above)
Comment (yurii-lubynets) [reply src/utils/profile.ts]: Now it makes sense. fixed in 5a3581bc4abfb14880e2e621a12c0994e01cace8

File: src/utils/profile.ts (hunk 22 above)
Comment (yurii-lubynets) [reply src/utils/profile.ts]: removed as unused. fixed in 5a3581bc4abfb14880e2e621a12c0994e01cace8

File: src/hooks/users.ts (hunk 23 above)
Comment (yurii-lubynets) [reply src/hooks/users.ts]: Actually, I could extract only `venueName` from user

```
const venueName = Object.keys(recentUser.lastSeenIn)[0] ?? "";
```
And `createUrlSafeName` method is the right (IMHO) way to get venueUrl. It's used to create venueUrl from venueName in the relevant form

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 24 above)
Comment (yurii-lubynets) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: Can't find anything useful. Please provide me an example

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx (hunk 25 above)
Comment (yurii-lubynets) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx]: reverted in 5a3581b
